types/resource.go

```diff
@@ -1,20 +1,282 @@
 package types
 
+import (
+	"reflect"
+)
+
+type UnitType int
+
+type Unit interface {
+	String() string
+	equals(Unit) bool
+	add(Unit) error
+	sub(Unit) error
+	le(Unit) bool
+}
+
+type ResUnit interface {
+	Equals(ResUnit) bool
+	Add(unit ResUnit) bool
+}
+
+// Resources stores Unit details and Count value
+type Resources struct {
+	Resources ResourceUnits `json:"resources"`
+	Count     uint32        `json:"count"`
+}
+
 // ResourceGroup is the interface that wraps GetName and GetResources methods
 type ResourceGroup interface {
 	GetName() string
-	GetResources() []Resource
+	GetResources() []Resources
 }
 
-// Resource stores Unit details and Count value
-type Resource struct {
-	Unit  Unit
-	Count uint32
+var _ Unit = (*CPU)(nil)
+var _ Unit = (*Memory)(nil)
+var _ Unit = (*Storage)(nil)
+
+// AddUnit it rather searches for existing entry of the same type and sums values
+// if type not found it appends
+func (m ResourceUnits) Add(rhs ResourceUnits) (res ResourceUnits, err error) {
+	res = m
+
+	defer func() {
```

Comment:
let's not use `defer` for this - just return like normal, without using `(res Foo, err error)` style variable declarations.